text/0000-generic-int.md

```diff
@@ -0,0 +1,665 @@
+00...000...01 (1)
+00...000...10 (2)
+...
+00...001...11 (2.pow(N) - 1)
+```
+
+Note that the next value sets our sign bit to one, but leaves the remaining
+values past the sign at zero. This starts the niche space, where the sign bit
+isn't the same as all the bits after it:
+
+```
+00...010...00 (2.pow(N) + 0)
+00...010...01 (2.pow(N) + 1)
+00...010...10 (2.pow(N) + 2)
+...
+00...011...11 (2.pow(N + 1) - 1)
+```
+
+Note that now, once we add one, our sign bit will flip back. However, even if we
+keep counting, it won't be extended all the way to the end. So, really, our
+niche extends quite far, until...
+
+```
+00...100...00 (+2.pow(N + 1))
+00...100...01 (+2.pow(N + 1) + 1)
+00...100...10 (+2.pow(N + 1) + 2)
+...
+11...101...11 (-2.pow(N) + 1)
+```
+
+Finally, at this point, adding one will flip our sign bit to be negative, and it
+will finally be extended all the way. After this, our ordinary values resume,
+starting at `-2.pow(N)` and ending at `-1`.
+
+As you can see, we have exactly the same number of values to use as niche, and
+they're all in one contiguous range.
+
+## Computations and storage
+
+### Overflow and underflow
+
+Because sign extension will always be applied, it's safe for the compiler to
+internally treat `uint<N>` as `uint<N.next_power_of_two()>` when doing
+computations. As a concrete example, this means that adding two `uint<48>`
+values will work exactly like adding two `u64` values, generating exactly the
+same code on targets with 64-bit registers. Right?
+
+Actually, it turns out that things will be slightly more complicated. In debug
+mode, the generated code will be very similar; overflow checks will still panic
+as usual. However, in release mode, as we saw in the above section, there will
+have to be additional code to make sure that overflows and underflows extend the
+sign correctly.
+
+This is actually a good case for adding a `add_unchecked` (and similar) methods
+which require code to ensure that overflow does not occur, skipping the extra
+code in release builds. Such methods would require their own RFC, but this
+actually does provide a decent reason why they'd be useful.
+
+Unfortunately, these generalisations aren't going to be free in all cases.
+
+### Storage sizes
+
+The restriction that `N <= 128` may allow for future changes where `uint<N>` is
+not actually stored as `uint<N.next_power_of_two()>` for sufficiently large `N`.
+Right now, even storing `uint<65>` as `u128` makes the most sense, because it
+will be the most efficient representation on 64-bit CPUs for performing many
+operations. However, storing `uint<4097>` as `uint<8192>`, for example, may be
+total overkill.
+
+The main goal for these representations is to avoid bit-wrangling as much as
+possible. Although truncating the number of bits for integers is no longer free,
+e.g. `u64` to `uint<48>` requires an AND and `i64` to `int<48>` requires
+conditionally excecuting an AND or an OR, the goal is that in most cases, using
```

Comment:
You can implement sign extension using a wrapping shift left, then a signed shift right. That is generally faster than conditionally executing a bitwise-and or a bitwise-or. Additionally, it's the canonical code sequence for sign-extension on RISC-V.